arraycontext/impl/pytato/__init__.py

```diff
@@ -207,23 +263,22 @@ def transform_dag(self, dag: "pytato.DictOfNamedArrays"
         return dag
 
     def tag(self, tags: Union[Sequence[Tag], Tag], array):
-        return array.tagged(tags)
+        return rec_map_array_container(lambda x: x.tagged(tags),
+                                       array)
 
     def tag_axis(self, iaxis, tags: Union[Sequence[Tag], Tag], array):
-        # TODO
-        from warnings import warn
-        warn("tagging PytatoPyOpenCLArrayContext's array axes: not yet implemented",
-             stacklevel=2)
-        return array
+        return rec_map_array_container(lambda x: x.with_tagged_axis(iaxis,
+                                                                     tags),
+                                        array)
 
     def einsum(self, spec, *args, arg_names=None, tagged=()):
-        import pyopencl.array as cla
+        from arraycontext.impl.pyopencl.taggable_cl_array import TaggableCLArray
         import pytato as pt
         if arg_names is None:
             arg_names = (None,) * len(args)
 
         def preprocess_arg(name, arg):
-            if isinstance(arg, cla.Array):
+            if isinstance(arg, TaggableCLArray):
```

Comment:
If you'd like to disallow passing in `cla.Array` here, raise an appropriate error instead of falling through.

Reply:
Yep, added a warning here too.